src/bin/mdbook-xgettext.rs

```diff
@@ -81,21 +90,242 @@ fn create_catalog(ctx: &RenderContext) -> anyhow::Result<Catalog> {
         add_message(&mut catalog, line, &source);
     }
 
+    let mut catalogs = HashMap::new();
+    let default_file_path = PathBuf::from(input_path.to_string());
+    // The catalog from the summary data will exist in the single pot file for
+    // a depth of 0, or be presented in a separate `SUMMARY.pot` file for
+    // depths greater than 0.
+    if depth == 0 {
+        catalogs.insert(default_file_path.display().to_string(), catalog);
+    } else {
+        catalogs.insert(summary_path.display().to_string(), catalog);
+    }
+
     // Next, we add the chapter contents.
-    for item in ctx.book.iter() {
+    for item in &ctx.book.sections {
         if let BookItem::Chapter(chapter) = item {
             let path = match &chapter.path {
                 Some(path) => ctx.config.book.src.join(path),
                 None => continue,
             };
+
+            let rel_file_path: String = if depth > 0 {
+                catalogs.insert(
+                    path.display().to_string(),
+                    Catalog::new(generate_catalog_metadata(&ctx)),
+                );
+                path.display().to_string()
+            } else {
+                default_file_path.display().to_string()
+            };
+
+            let catalog: &mut Catalog = catalogs
+                .get_mut(&rel_file_path)
+                .with_context(|| anyhow!("Expected to find {:?} in map", rel_file_path))?;
+
             for (lineno, msgid) in extract_messages(&chapter.content) {
                 let source = format!("{}:{}", path.display(), lineno);
-                add_message(&mut catalog, &msgid, &source);
+                add_message(catalog, &msgid, &source);
             }
+
+            let file_path = if depth == 0 {
+                Some(default_file_path.display().to_string())
+            } else if depth == 1 {
+                Some(chapter.path.clone().unwrap().display().to_string())
+            } else {
+                None
+            };
+            get_subcontent_for_chapter(chapter.clone(), file_path, depth)
+                .iter()
+                .for_each(
```

Comment:
Please try converting this to a regular `for` loop — using `for_each` is almost never a nice syntax in my opinion. The extra indentation here is a sign that too much stuff is happening at once :slightly_smiling_face: